drivers/ethernet/phy/phy_intel_igc.c

```diff
@@ -0,0 +1,371 @@
+	} else if ((anar_val & anlpar_val) & MII_ADVERTISE_100_HALF) {
+		state->speed = LINK_HALF_100BASE;
+	} else if ((anar_val & anlpar_val) & MII_ADVERTISE_10_FULL) {
+		state->speed = LINK_FULL_10BASE;
+	} else {
+		state->speed = LINK_HALF_10BASE;
+	}
+
+	return ret;
+}
+
+static int phy_intel_igc_get_link_state(const struct device *dev,
+					struct phy_link_state *state)
+{
+	uint32_t value = 0;
+	int ret;
+
+	ret = phy_intel_igc_await_phy_status(dev, MII_BMSR, MII_BMSR_LINK_STATUS,
+					     &value, PHY_IGC_READ_DELAY);
+	if (ret < 0) {
+		return ret;
+	}
+
+	state->is_up = !!(value & MII_BMSR_LINK_STATUS);
+
+	return ret;
+}
+
+static void invoke_link_cb(const struct device *dev)
+{
+	struct phy_intel_igc_data *const data = dev->data;
+
+	if (!data->cb) {
+		LOG_INF("Callback not yet registered");
+		return;
+	}
+
+	data->cb(data->phy, &data->state, data->cb_data);
+}
+
+static int phy_intel_igc_link_cb_set(const struct device *dev,
+				     phy_callback_t cb, void *user_data)
+{
+	struct phy_intel_igc_data *data = dev->data;
+
+	data->cb = cb;
+	data->cb_data = user_data;
+
+	invoke_link_cb(dev);
+
+	return 0;
+}
+
+static void update_link_state(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct phy_intel_igc_data *data = CONTAINER_OF(dwork, struct phy_intel_igc_data,
+						       link_work);
+	const struct device *dev = data->phy;
+	struct phy_link_state new_state = {0};
+
+	if (phy_intel_igc_get_link_state(dev, &new_state) < 0) {
+		LOG_ERR("PHY Link is Down");
+	}
+
+	if (phy_intel_igc_get_speed(dev, &new_state) < 0) {
+		LOG_ERR("Failed to get PHY link Speed");
+	}
+
+	if (new_state.is_up != data->state.is_up || new_state.speed != data->state.speed) {
+		data->state.is_up = new_state.is_up;
+		data->state.speed = new_state.speed;
+
+		invoke_link_cb(dev);
+
+		if (data->state.is_up) {
+			LOG_INF("PHY Link is Up with Speed %s Mbps & %s duplex",
+				PHY_LINK_IS_SPEED_1000M(new_state.speed) ? "1000" :
+				(PHY_LINK_IS_SPEED_100M(new_state.speed) ? "100" : "10"),
+				PHY_LINK_IS_FULL_DUPLEX(new_state.speed) ? "Full" : "Half");
+		}
+	}
+}
+
+void eth_intel_igc_misc_phy_link_isr(const struct device *dev)
+{
+	struct phy_intel_igc_data *data = dev->data;
+
+	k_work_schedule(&data->link_work, K_MSEC(0));
+}
+
+static int phy_intel_igc_init(const struct device *dev)
+{
+	struct phy_intel_igc_data *data = dev->data;
+	int ret;
+
+	data->phy = dev;
+	ret = phy_intel_igc_id_read(dev);
+	if (ret < 0) {
+		LOG_ERR("Failed to read PHY ID");
+		return ret;
+	}
+
+	ret = phy_intel_igc_aneg_10_100mbps(dev);
+	if (ret < 0) {
+		LOG_ERR("Autoneg init failed for 10/100 Mbps");
+		return ret;
+	}
+
+	ret = phy_intel_igc_aneg_1000mbps(dev);
+	if (ret < 0) {
+		LOG_ERR("Autoneg init failed for 1000 Mbps");
+		return ret;
+	}
+
+	ret = phy_intel_igc_restart_aneg(dev);
+	if (ret < 0) {
+		LOG_ERR("Failed to restart Autoneg");
+	}
+
+	k_work_init_delayable(&data->link_work, update_link_state);
```

Comment:
this should be init before the interrupt is activated. it would also be good to schedule a run at the end of this function, so we get the current state, as the phy could already be up. maybe use `Z_WORK_DELAYABLE_INITIALIZER` in `INTEL_IGC_PHY_INIT`